File: maps/src/main/java/it/localhostsoftware/maps/huawei/model/HuaweiPolylineOptions.java
package it.localhostsoftware.maps.huawei.model;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import it.localhostsoftware.maps.model.LatLng;
import it.localhostsoftware.maps.model.PatternItem;
import it.localhostsoftware.maps.model.PolylineOptions;

public class HuaweiPolylineOptions extends PolylineOptions<com.huawei.hms.maps.model.PolylineOptions> {
    public HuaweiPolylineOptions(com.huawei.hms.maps.model.PolylineOptions polylineOptions) {
        super(polylineOptions);
    }

    @Override
    public PolylineOptions<?> add(LatLng<?> var1) {
        getPolylineOptions().add((com.huawei.hms.maps.model.LatLng) var1.getLatLng());
        return this;
    }

    @Override
    public PolylineOptions<?> add(LatLng<?>... var1) {
        for (LatLng<?> value : var1)
            add(value);
        return this;
    }

    @Override
    public PolylineOptions<?> addAll(Iterable<LatLng<?>> var1) {
        for (LatLng<?> value : var1)
            add(value);
        return this;
    }

    @Override
    public PolylineOptions<?> width(float var1) {
        getPolylineOptions().width(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> color(int var1) {
        getPolylineOptions().color(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> jointType(int var1) {
        getPolylineOptions().jointType(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> pattern(@Nullable List<PatternItem<?>> var1) {
        if (var1 == null)
            getPolylineOptions().pattern(null);
        else {
            ArrayList<com.huawei.hms.maps.model.PatternItem> values = new ArrayList<>(var1.size());
            for (PatternItem<?> value : var1)
                values.add((com.huawei.hms.maps.model.PatternItem) value.getPatternItem());
            getPolylineOptions().pattern(values);
        }
        return this;
    }

    @Override
    public PolylineOptions<?> zIndex(float var1) {
        getPolylineOptions().zIndex(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> visible(boolean var1) {
        getPolylineOptions().visible(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> geodesic(boolean var1) {
        getPolylineOptions().geodesic(var1);
        return this;
    }

    @Override
    public PolylineOptions<?> clickable(boolean var1) {
        getPolylineOptions().clickable(var1);
        return this;
    }

    @Override
    public List<LatLng<?>> getPoints() {
        ArrayList<LatLng<?>> out = new ArrayList<>(getPolylineOptions().getPoints().size());
        for (com.huawei.hms.maps.model.LatLng latLng : getPolylineOptions().getPoints())
            out.add(new HuaweiLatLng(latLng));
        return out;
    }

    @Override
    public float getWidth() {
        return getPolylineOptions().getWidth();
    }

    @Override
    public int getColor() {
        return getPolylineOptions().getColor();
    }

    @Override
    public int getJointType() {
        return getPolylineOptions().getJointType();
    }

    @Nullable
    @Override
    public List<PatternItem<?>> getPattern() {
        if (getPolylineOptions().getPattern() == null)
            return null;
        else {
            ArrayList<PatternItem<?>> out = new ArrayList<>(getPolylineOptions().getPattern().size());
            for (com.huawei.hms.maps.model.PatternItem value : getPolylineOptions().getPattern())
                out.add(new HuaweiPatternItem(value));
            return out;
        }
    }

    @Override
    public float getZIndex() {
        return getPolylineOptions().getZIndex();
    }

    @Override
    public boolean isVisible() {
        return getPolylineOptions().isVisible();
    }

    @Override
    public boolean isGeodesic() {
        return getPolylineOptions().isGeodesic();
    }

    @Override
    public boolean isClickable() {
        return getPolylineOptions().isClickable();
    }
}

File: maps/src/main/java/it/localhostsoftware/maps/huawei/model/HuaweiPolylineOptions.kt
package it.localhostsoftware.maps.huawei.model

import com.huawei.hms.maps.model.PolylineOptions

class HuaweiPolylineOptions(polylineOptions: PolylineOptions) : it.localhostsoftware.maps.model.PolylineOptions<PolylineOptions, HuaweiPatternItem, HuaweiLatLng, HuaweiCap>(polylineOptions) {
    override fun add(vararg var1: HuaweiLatLng): it.localhostsoftware.maps.model.PolylineOptions<PolylineOptions, HuaweiPatternItem, HuaweiLatLng, HuaweiCap> {
        po.add(*var1.map { it.ll }.toTypedArray())
        return this
    }

    override val points: List<HuaweiLatLng>
        get() = po.points.map { HuaweiLatLng(it) }
    override var width: Float
        get() = po.width
        set(value) {
            po.width(value)
        }
    override var color: Int
        get() = po.color
        set(value) {
            po.color(value)
        }
    override var startCap: HuaweiCap
        get() = HuaweiCap(po.startCap)
        set(value) {
            po.startCap(value.cap)
        }
    override var endCap: HuaweiCap
        get() = HuaweiCap(po.endCap)
        set(value) {
            po.endCap(value.cap)
        }
    override var jointType: Int
        get() = po.jointType
        set(value) {
            po.jointType(value)
        }
    override var pattern: List<HuaweiPatternItem>?
        get() = po.pattern?.map { HuaweiPatternItem(it) }
        set(value) {
            po.pattern(value?.map { it.pi })
        }
    override var zIndex: Float
        get() = po.zIndex
        set(value) {
            po.zIndex(value)
        }
    override var isVisible: Boolean
        get() = po.isVisible
        set(value) {
            po.visible(value)
        }
    override var isGeodesic: Boolean
        get() = po.isGeodesic
        set(value) {
            po.geodesic(value)
        }
    override var isClickable: Boolean
        get() = po.isClickable
        set(value) {
            po.clickable(value)
        }
}